ESQL: Simplify TableIdentifier class + rename to IndexPattern

alex-spies opened the pull request:
This class is confusing:
- It contains an **unused** `cluster` attribute - we never separate out the cluster, it remains in the `index` field. Also, in the constructor, this field is called `catalog`, which is a concept entirely absent from ESQL at the moment.
- It can refer to multiple indices, even multiple wildcard patterns, but doesn't mention this neither in its name nor javadoc.
- It has little to do with tables, which is likely a remnant of this class' usage in SQL, before the `esql.core` split.

This PR removes the `cluster` attribute, renames the class to `IndexPattern`, and adds javadoc to clarify that it can also contain stuff like `remote1:idx1,remote-*:idx-*`.

alex-spies committed:
Remove TableIdentifier.cluster()

This was always null throughout the entire code base.

alex-spies committed:
Rename and add javadoc

alex-spies committed:
Actually, rename to just IndexPattern

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/TableInfo.java:
```diff
@@ -7,17 +7,17 @@
-import org.elasticsearch.xpack.esql.plan.TableIdentifier;
+import org.elasticsearch.xpack.esql.plan.IndexPattern;
 
 public class TableInfo {
```
In a future PR, I think this class just has to go.

idegtiarenko commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/UnresolvedRelation.java:
```diff
@@ -71,7 +71,7 @@ protected NodeInfo<UnresolvedRelation> info() {
     }
 
-    public TableIdentifier table() {
+    public IndexPattern table() {
```
I wonder if the method should be renamed as well? Or may be it is used in too many places

alex-spies replied:
Yeah, that name is just misleading at the moment. But I'd like to do this in another PR to keep the blast radius contained.

alex-spies replied:
Actually, I changed my mind. Let's go and have a little more readability and consistency.

idegtiarenko commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/IndexPattern.java:
```diff
@@ -0,0 +1,57 @@
+public class IndexPattern {
+
+    private final Source source;
+    private final String indexPattern;
```
I wonder if this should be called indexPattern(s) as it might list coma separated patterns?

Not sure if it is needed often, but https://github.com/elastic/elasticsearch/pull/120277 would benefit if it is possible to return list/array of separate patterns.

alex-spies replied:
We can add a helper method as part of https://github.com/elastic/elasticsearch/pull/120277! I'd like to not add logic in this PR, so it remains a straight refactoring.

alex-spies replied:
Also, yeah, `indexPatterns` (plural) would be even more precise - but then we'd have to align e.g. `EsRelation` as well. And it's also reasonable to consider `idx,logs-*` a single pattern which matches `idx` and anything starting with `logs-`.

alex-spies committed:
Better method names

Rename IndexPattern.index to indexPattern and UnresolvedRelation.table
to indexPattern.

alex-spies committed:
Merge branch 'main' into simplify_table_class

elasticsearchmachine commented:
## 💔 Backport failed
| Status | Branch | Result |
|:------:|:------:|:------:|
| ❌ |  8.x  | Commit could not be cherrypicked due to conflicts |

You can use [sqren/backport](https://github.com/sqren/backport) to manually backport by running `backport --upstream elastic/elasticsearch --pr 120797`
